Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java

```diff
@@ -0,0 +1,131 @@
+ */
+public final class OsgirTheReconstructor extends CardImpl {
+
+    public OsgirTheReconstructor(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{2}{R}{W}");
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.GIANT);
+        this.subtype.add(SubType.ARTIFICER);
+        this.power = new MageInt(4);
+        this.toughness = new MageInt(4);
+
+        // Vigilance
+        this.addAbility(VigilanceAbility.getInstance());
+
+        // {1}, Sacrifice an artifact: Target creature you control gets +2/+0 until end of turn.
+        Ability ability = new SimpleActivatedAbility(Zone.BATTLEFIELD, new BoostTargetEffect(2, 0, Duration.EndOfTurn), new ManaCostsImpl<>("{1}"));
+        ability.addTarget(new TargetCreaturePermanent());
+        ability.addCost(new SacrificeTargetCost(new TargetControlledPermanent(new FilterControlledArtifactPermanent("an artifact"))));
+        this.addAbility(ability);
+
+        // {X},{T}, Exile an artifact with mana value X from your graveyard: Create two tokens that are copies of the exiled card. Activate only as
+        Ability copyAbility = new ActivateAsSorceryActivatedAbility(Zone.BATTLEFIELD,
+                new OsgirTheReconstructorCreateArtifactTokensEffect(),
+                new ManaCostsImpl("{X}"));
+        copyAbility.addCost(new TapSourceCost());
+        copyAbility.setTargetAdjuster(OsgirTheReconstructorAdjuster.instance);
+
+        this.addAbility(copyAbility);
+    }
+
+    private OsgirTheReconstructor(final OsgirTheReconstructor card) {
+        super(card);
+    }
+
+    @Override
+    public OsgirTheReconstructor copy() {
+        return new OsgirTheReconstructor(this);
+    }
+}
+
+enum OsgirTheReconstructorAdjuster implements TargetAdjuster {
+    instance;
+
+    @Override
+    public void adjustTargets(Ability ability, Game game) {
+        int xValue = ability.getManaCostsToPay().getX();
+        FilterCard filter = new FilterArtifactCard("an artifact card with mana value "+xValue+" from your graveyard");
+        filter.add(new ManaValuePredicate(ComparisonType.EQUAL_TO, xValue));
+        ability.getTargets().clear();
+        ability.getTargets().add(new TargetCardInYourGraveyard(filter));
+    }
+}
+
+class OsgirTheReconstructorCreateArtifactTokensEffect extends OneShotEffect {
+
+    public OsgirTheReconstructorCreateArtifactTokensEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Create two tokens that are copies of the exiled card.";
+    }
+
+    public OsgirTheReconstructorCreateArtifactTokensEffect(final OsgirTheReconstructorCreateArtifactTokensEffect effect)  {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        Card card = game.getCard(targetPointer.getFirst(game, source));
+
+        player.moveCards(card, Zone.EXILED, source, game);
```

Comment:
The card needs to be exiled as part of the cost, not the effect. You need to created a `CostAdjuster` to go with the `TargetAdjuster`. Take a look at cards like Necropolis Fiend and Scorched Earth.

Reply:
everytime I try to add the exiled card as a cost, I cannot seem to be able to get it to copy. changed adjustTargets to this:

public void adjustTargets(Ability ability, Game game) {
        int xValue = ability.getManaCostsToPay().getX();
        FilterCard filter = new FilterArtifactCard("an artifact card with mana value "+xValue+" from your graveyard");
        filter.add(new ManaValuePredicate(ComparisonType.EQUAL_TO, xValue));

        ability.getTargets().clear();
        for (Cost cost : ability.getCosts()) {
            if (!(cost instanceof ExileFromGraveCost)) {
                continue;
            }
            ExileFromGraveCost exileCost = (ExileFromGraveCost) cost;
            for (Target target : exileCost.getTargets()) {
                if (target instanceof TargetCardInYourGraveyard) {
                    ability.getTargets().add(target);
                }
            }
        }
    }

but it does not target anything

also, is there a way to set breakpoints inside Mage.Sets? The debugger seems to skip righ over those

Reply:
Oh I just realized that you don't need a `TargetAdjuster` at all, you just need a `CostAdjuster`. You might actually want to create a custom cost instead like Marath, Will of the Wild has. 

As for breakpoints, I'm a shitty dev who uses print statements so I can't answer that.

Reply:
haha, alright thanks

Reply:
ok, I ended up being able fix it with just a cost adjuster. submitted all changes